# **ECON3203 - Group DFM - Implementation File Part 2**

<font color = purple> Faiyam Islam (z5258151) </font> 

<font color = purple> Mimi Nguyen (z5260968) </font>

<font color = purple> Dharani Palanisamy (z5260276) </font>

<font color = purple> Dharshini Loganathan (z5309765) </font>

This implementation file contains: 

1. K-Nearest Neighbours Algorithm
2. Neural Networks and Deep Learning

Due to the time-consumption of KNN and Neural Networks, we have decided to implement those methods on this file. 

The test_error can be found at the bottom of this notebook. The final chosen for our group is the deep learning (neural networks) model. 

# **K-Nearest Neighbours Algorithm**


The KNN method has been implemented here as a supervised learning algorithm to be used for regression problems. Various test errors have been achieved here, the lowest will be selected. 

In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns
import statsmodels.formula.api as smf 
import warnings 
warnings.filterwarnings('ignore')
%matplotlib inline
sns.set_context('notebook')
sns.set_style('ticks')

from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

In [ ]:
train = pd.read_csv('ATM_training.csv')
ATM_test = pd.read_csv('ATM_test.csv')

In [ ]:
# Standardising the data
response = ['Withdraw']
numerical_predictors=['Shops', 'ATMs']
predictors=[x for x in list(train.columns) if x not in response]

mu=train[numerical_predictors].mean()
sigma=train[numerical_predictors].std()
train[numerical_predictors]=(train[numerical_predictors]-mu)/sigma
ATM_test[numerical_predictors]=(ATM_test[numerical_predictors]-mu)/sigma

In [ ]:
# Function
def knn_test(predictors, response):
    ## Function inputs are predictor and repsonse variables.
    ## It will try out many different values of k & keep track of 
    ## the optimal k which gives the best CV score & the lowest MSE.
    ## With the final optimal value of k, function will then fit the 
    ## training data, then make predictions using VALIDATION(?) data. 
    ## Function outputs are val_mse, cv_mse.

    neighbours=np.arange(1, 51) 
    best_score = -np.inf
    
    for k in neighbours: 
        knn = KNeighborsRegressor(n_neighbors = k, metric='mahalanobis', metric_params={'V': train[predictors].cov()}) 
        scores = cross_val_score(knn, train[predictors], train[response], cv=10, scoring = 'neg_mean_squared_error')
        cv_score = np.mean(scores)
        if cv_score >= best_score: 
            best_score = cv_score 
            best_knn = knn 

    ## training knn model with optimal k and training data 
    knn = best_knn
    knn.fit(train[predictors], train[response])
    predictions = knn.predict(test[predictors])
    test_mse = mean_squared_error(test[response], predictions)
   
    cv_mse= np.sqrt(-best_score)
    print('Chosen K: {}'.format(knn.n_neighbors))
    
    return test_mse, cv_mse

In [ ]:
# Maybe we should remove this and the one below this cell 
predictors = ['Shops']
test_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(test_mse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4)))

Chosen K: 50
Test MSE : 3.52
Cross validation MSE : 4.1638


In [ ]:
predictors = ['ATMs']
test_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(test_mse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4)))

Chosen K: 48
Test MSE : 4.0575
Cross validation MSE : 6.9434


In [ ]:
predictors = ['Shops', 'ATMs']
test_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(test_mse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4)))

Chosen K: 49
Test MSE : 2.6253
Cross validation MSE : 3.7379


In [ ]:
predictors = ['Shops', 'ATMs', 'Downtown']
test_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(test_mse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4)))

Chosen K: 49
Test MSE : 2.8323
Cross validation MSE : 3.7316


In [ ]:
predictors = ['Shops', 'ATMs', 'Weekday'] 
test_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(testmse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4)))

Chosen K: 41
Test MSE : 1.905
Cross validation MSE : 3.3625


In [ ]:
predictors = ['Shops', 'ATMs', 'Center'] 
test_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(test_mse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4)))

Chosen K: 44
Test MSE : 2.7453
Cross validation MSE : 2.9433


In [ ]:
predictors = ['Shops', 'ATMs', 'High'] 
test_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(test_mse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4)))

Chosen K: 46
Test MSE : 2.8375
Cross validation MSE : 3.7245


Based on the results above, we observe that including Weekday and Center can improve our knn model whilse High and Downtown have minimal effects. 

We further test out various predictor combinations of Shops, ATMs, Weekday and Center, and work our way towards the predictor subset chosen by the LASSO method. 

In [ ]:
predictors = ['Shops', 'ATMs', 'Weekday', 'Center']
test_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(test_mse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4))) 

Chosen K: 9
Test MSE : 0.3615
Cross validation MSE : 0.7521


In [ ]:
## Corresponds to LASSO regression results
predictors = ['Shops', 'ATMs', 'Weekday', 'Center', 'High']
test_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(test_mse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4))) #lowest so far

Chosen K: 5
Test MSE : 0.2567
Cross validation MSE : 0.6101


In [ ]:
predictors = ['Shops', 'ATMs', 'Center', 'High']
test_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(test_mse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4))) 

Chosen K: 50
Test MSE : 3.2765
Cross validation MSE : 2.9224


In [ ]:
predictors = ['Shops', 'ATMs', 'Center', 'High', 'Downtown']
test_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(test_mse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4))) 

Chosen K: 50
Test MSE : 3.1326
Cross validation MSE : 2.9018


In [ ]:
predictors = ['Shops', 'Center', 'High', 'Downtown'] 
test_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(test_mse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4))) 

Chosen K: 47
Test MSE : 4.5329
Cross validation MSE : 3.4071


In [ ]:
predictors = ['Shops', 'ATMs', 'Downtown', 'Weekday', 'Center']
test_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(test_mse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4))) 

Chosen K: 16
Test MSE : 0.3456
Cross validation MSE : 0.7153


In [ ]:
predictors = ['Shops', 'ATMs', 'Downtown', 'Weekday', 'Center', 'High']
val_mse, cv_mse  =  knn_test(predictors, response) 
print('Test MSE : {}'.format(test_mse.round(4)))
print('Cross validation MSE : {}'.format(cv_mse.round(4)))

Chosen K: 12
Validation MSE : 0.3456
Cross validation MSE : 0.7153


**By inspection of the KNN Regression results, we obtain a test mean squared of error of 0.3456 being the smallest in comparison to all the other methods used here.**

# **Neural Networks and Deep Learning**




In [ ]:
import math
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_squared_error

from keras.layers.core import Dense 
from keras.models import Sequential
import tensorflow as tf

In [ ]:
np.random.seed(3)

In [ ]:
tf.random.set_seed(3)

In [ ]:
# Load the data
df = pd.read_csv('ATM_training.csv')
ATM_test = pd.read_csv('ATM_test.csv')
df.head()

Shops  ATMs  Downtown  Weekday  Center  High   Withdraw
0  10.18    10         1        0       0     0  72.750556
1   9.74    10         1        1       0     0  66.720482
2   0.96     2         0        0       0     1  19.189516
3   9.58     9         1        1       0     1  67.388669
4   1.03     4         0        1       0     1  15.813127

In [ ]:
#Standardising predictors for training set

response = ['Withdraw']
numerical_predictors = ['Shops','ATMs'] # the list all predictors

mu = df[numerical_predictors].mean() # mean for each feature
sigma = df[numerical_predictors].std() # std for each feature

df[numerical_predictors] = (df[numerical_predictors] - mu) / sigma

In [ ]:
df.head()

Shops      ATMs  Downtown  Weekday  Center  High   Withdraw
0  0.695276  0.561479         1        0       0     0  72.750556
1  0.588446  0.561479         1        1       0     0  66.720482
2 -1.543299 -1.616331         0        0       0     1  19.189516
3  0.549598  0.289253         1        1       0     1  67.388669
4 -1.526303 -1.071879         0        1       0     1  15.813127

In [ ]:
# Splitting into train and validation data - using time window
train_size = math.ceil(0.7*len(df))

train = df.sample(n=train_size, random_state=1)
val = df[df.index.isin(train.index)==False]

In [ ]:
df = df.values  # Convert from DataFrame to Python Array
train = train.values  
val = val.values

In [ ]:
df = df.astype('float') # Data is set as type 'float'
train = train.astype('float') 
val = val.astype('float')

In [ ]:
# Split the dataset into input (X) and output variables (y)
X = df[:,0:6]
y = df[:,6]

# training data 
X_train = train[:,0:6]
y_train = train[:,6]

# validation data 
X_val = val[:,0:6]
y_val = val[:,6]

In [ ]:
model = Sequential()
model.add(Dense(12, input_dim = 6, activation ='relu'))
model.add(Dense(1))

In [ ]:
model.compile(loss='mean_squared_error', optimizer='adam')

In [ ]:
model.summary()


Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_2 (Dense)             (None, 12)                84        
                                                                 
 dense_3 (Dense)             (None, 1)                 13        
                                                                 
Total params: 97
Trainable params: 97
Non-trainable params: 0
_________________________________________________________________


In [ ]:
model.fit(X_train, y_train, epochs = 150, batch_size = 32 , verbose = 1)

Epoch 1/150
482/482 [==============================] - 1s 2ms/step - loss: 3294.1663
Epoch 2/150
482/482 [==============================] - 1s 1ms/step - loss: 1632.7827
Epoch 3/150
482/482 [==============================] - 1s 1ms/step - loss: 373.3582
Epoch 4/150
482/482 [==============================] - 1s 1ms/step - loss: 139.4779
Epoch 5/150
482/482 [==============================] - 1s 1ms/step - loss: 106.6469
Epoch 6/150
482/482 [==============================] - 1s 1ms/step - loss: 82.4684
Epoch 7/150
482/482 [==============================] - 1s 1ms/step - loss: 59.7028
Epoch 8/150
482/482 [==============================] - 1s 1ms/step - loss: 38.7006
Epoch 9/150
482/482 [==============================] - 1s 1ms/step - loss: 21.7861
Epoch 10/150
482/482 [==============================] - 1s 1ms/step - loss: 11.7209
Epoch 11/150
482/482 [==============================] - 1s 1ms/step - loss: 8.1284
Epoch 12/150
482/482 [==============================] - 1s 1ms/step - loss: 7.3

In [ ]:
allPredict = model.predict(X)

In [ ]:
trainScore_not_scaled = (mean_squared_error(y[:train_size], allPredict[:train_size, 0]))
print('Training Data MSE on original scale: {0:.4f}'.format(trainScore_not_scaled))

Training Data MSE on original scale: 0.2613


In [ ]:
valScore_not_scaled = (mean_squared_error(y[train_size:], allPredict[train_size:,0]))
print('Validation Data MSE original scale: {0:.4f}'.format(valScore_not_scaled))

Validation Data MSE original scale: 0.2550


In [ ]:
#Standardising predictors for test set

response = ['Withdraw']
numerical_predictors = ['Shops','ATMs'] # the list all predictors

# mu = ATM_test[numerical_predictors].mean() # mean for each feature
# sigma = ATM_test[numerical_predictors].std() # std for each feature

ATM_test[numerical_predictors] = (ATM_test[numerical_predictors] - mu) / sigma

# Converting test set values into Python array
ATM_test = ATM_test.values  

# Data is set as type 'float'
ATM_test = ATM_test.astype('float') 


In [ ]:
# Split the dataset into input (X) and output variables (y)
X_test = ATM_test[:,0:6]
y_test = ATM_test[:,6]

# Predicting
allPredict = model.predict(X_test)

#Test MSE
test_error = (mean_squared_error(y_test, allPredict))
print('Test Data MSE on original scale: {0:.4f}'.format(test_error))

Test Data MSE on original scale: 0.2432


# **Evaluation of analytical methods**


|SLR||MLR||Ridge||LASSO||Enet||KNN||NN|
|---||---||-----||-----||----||---||---|
|17.7686||0.3108||6.2583||6.7418||6.7451||0.6101||0.2550|

<font color = purple>**Based off the analytical methods utilised, the lowest mean squared error achieved is from the neural networks model. Thus our selected cross validation mean squared error from all the other methods is 0.2550**</font>

In [ ]:
print(test_error)

0.243154618306369
